tratto/src/main/java/star/tratto/util/javaparser/JavaParserUtils.java

```diff
@@ -0,0 +1,1078 @@
+     * but not the other way around. Note that this method takes as input pairs of package and class
+     * name, instead of fully qualified types.
+     * @param type1 pair with &lt;package, class&gt;, e.g., &lt;"java.util", "List"&gt;
+     * @param type2 pair with &lt;package, class&gt;, e.g., &lt;"java.lang", "Object"&gt;
+     * @param oracleDatapoint may be null. If not null, it is used to check if some type is generic.
+     * @return true if a variable of type1 can be assigned to a variable of type2, false otherwise
+     */
+    public static boolean isType1AssignableToType2(Pair<String, String> type1, Pair<String, String> type2, OracleDatapoint oracleDatapoint) {
+        // Cases to consider:
+        // 1) left and right types are the same
+        // 2) left type is null and right type is not primitive
+        // 3) left type is primitive and right type is wrapper
+        // 4) both types are numeric and left type is assignable to right type
+        // 5) left type is instance of right type (both types are non-primitive)
+        return
+                type1.equals(type2) ||
+                (type1.equals(JavaTypes.NULL) && !JavaTypes.PRIMITIVES.contains(type2)) ||
+                (JavaTypes.PRIMITIVES.contains(type1) && JavaTypes.PRIMITIVES_TO_WRAPPERS.get(type1).equals(type2)) ||
+                (JavaTypes.NUMBERS.contains(type1) && JavaTypes.NUMBERS.contains(type2) && isAssignableToNumeric(type1, type2)) ||
+                isInstanceOf(fullyQualifiedClassName(type1), fullyQualifiedClassName(type2), oracleDatapoint);
+    }
+
+    public static TypeDeclaration<?> getClassOrInterface(CompilationUnit cu, String name) {
+        try {
+            return cu.getLocalDeclarationFromClassname(name).get(0);
+        } catch (NoSuchElementException|IndexOutOfBoundsException ignored) {}
+        try {
+            return cu.getClassByName(name).get();
+        } catch (NoSuchElementException ignored) {}
+        try {
+            return cu.getInterfaceByName(name).get();
+        } catch (NoSuchElementException ignored) {}
+        try {
+            return cu.getEnumByName(name).get();
+        } catch (NoSuchElementException e) {
+            throw new RuntimeException("Could not find class or interface " + name + " in compilation unit.", e);
+        }
+    }
+
+    public static TypeDeclaration<?> getClassOrInterface(String classSourceCode, String name) {
+        return getClassOrInterface(javaParser.parse(classSourceCode).getResult().get(), name);
+    }
+
+    /**
+     * Returns the String variable declaration represented by a JavaParser
+     * VariableDeclarator. Recall that a JavaParser field declaration may
+     * declare multiple variables. Each variable is separated into individual
+     * declarations.
+     *
+     * @param field the JavaParser field declaration
+     * @param variable the JavaParser variable declaration
```

Comment:
Why are both `field` and `variable` required?
What if the variable is not a field but (say) a local variable or a formal parameter?